### PR TITLE
Tutorial 4 Convert

### PR DESCRIPTION
**Pull request by F20170604**
Image Classification with Convolutional Neural Networks
Original File Link : https://github.com/tensorflow/examples/blob/master/courses/udacity_intro_to_tensorflow_for_deep_learning/l04c01_image_classification_with_cnns.ipynb

### CODE REVIEW

**Comment by BradLarson** [comment]
Because the initial parts of this are largely the same as for Tutorial 3, many of my style comments I just submitted on the lines there apply here. Those same changes could be made here, too.

**Comment by F20170604** [comment]
> Because the initial parts of this are largely the same as for Tutorial 3, many of my style comments I just submitted on the lines there apply here. Those same changes could be made here, too.

Yes yes. Most of the Tutorial 4 is same as 3. Only the network used and it's explanation is different. Else all is same. So whatever changes I make in 3, I'll make the same in 4 too.

**Comment by BradLarson** [comment]
In the source Tutorial 4, they use this sequencing for the model:

```
model = tf.keras.Sequential([
    tf.keras.layers.Conv2D(32, (3,3), padding='same', activation=tf.nn.relu,
                           input_shape=(28, 28, 1)),
    tf.keras.layers.MaxPooling2D((2, 2), strides=2),
    tf.keras.layers.Conv2D(64, (3,3), padding='same', activation=tf.nn.relu),
    tf.keras.layers.MaxPooling2D((2, 2), strides=2),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dense(10,  activation=tf.nn.softmax)
])
```

looking at the model in the ported version, I see

```
    var conv1a = Conv2D<Float>(filterShape: (3, 3, 1, 32), activation: relu)
    var conv1b = Conv2D<Float>(filterShape: (3, 3, 32, 64), activation: relu)
    var pool1 = MaxPool2D<Float>(poolSize: (2, 2), strides: (2, 2))
    var dropout1a = Dropout<Float>(probability: 0.25)
    var flatten = Flatten<Float>()
    var layer1a = Dense<Float>(inputSize: (3 * 3 * 32 * 64) / 2, outputSize: 128, activation: relu)
    var dropout1b = Dropout<Float>(probability: 0.5)
    var layer1b = Dense<Float>(inputSize: 128, outputSize: 10, activation: softmax)
```

I think there might be a MaxPool missing after the first convolution layer, and the original tutorial didn't seem to use dropout within it. Maybe the model architecture could be adjusted to bring it more closely in line with the original tutorial, for consistency.

**Comment by F20170604** [comment]
@BradLarson I've modified the notebook to be as same as Tutorial 3.
And I've made changes to the Classifier to be the exact same as that in Original Tutorial.

File: S4TF_Tutorial_4.ipynb [hunk 1]
```diff
@@ -0,0 +1,1807 @@
...
+        "colab_type": "code",
+        "outputId": "574c7b4a-14e9-4c54-e6b1-870cf8c17f55",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 281
+        }
+      },
+      "source": [
+        "let firstTrainImage = trainImages[0]\n",
+        "let firstTrainLabel = Int((trainNumericLabels[0].scalars)[0])\n",
```
Comment (rxwei) [inline S4TF_Tutorial_4.ipynb:532]:
```suggestion
        "let firstTrainLabel = Int(trainNumericLabels[0][0].scalarized())\n",
```

File: S4TF_Tutorial_4.ipynb [hunk 2]
```diff
@@ -0,0 +1,1807 @@
...
...Xdd9/FhAkTAKx9TPma0LVrV2u7lSz5msLHyq2KOHfuXGtfdtll1uYsBADQo0cP\naz/wwAPW/sMf/mDt/fbbL7FOqO6ykBZqwtc8Pg5Z9xOzVhiumgsAEydOtDbn5uZqwkAyawwfOzdD\nDo8R0o4/Z2fyHX8O9wD8oTkdDUEqwX3lMA83lNSXIYd9IbD298YlJp3E0Qul7mTJTygiw0XkZRGZ\nWywH6i4/V0RmiMhUEXlYRPx3MkpTkDWPpWolf6hPUUJRn6KEEItP0YFyDsmSdkVEOgG4GsAIADsA\nOF5EdnCaPQ+gzRizM4A7AFxag+4rdSJreh7VSv5Qn6KEkiU9nGolf8TkU3SgnFMy3KkNAzDXGDPP\nGPMxgFsBHMENjDGPGmNKKQUmAuhb9Y4rdSXjL8qqlRyiPkUJRX2KEkIsPiXKGOW0qnhMWgUYX3yw\nm27pnXfesTbHJYdWLGLc2GWOJ3z//fet7Vay4Vgd7rcbk+zGQWUlJfanu4hMptejjDGl/Dh9AHBZ\noQUAkjmpkpwG4B8VdVRpKBl1AkSileuvv97ad911l7X5uwiEp4XMQlosa73gOMu33347sYzjtdlP\n7r333h3aR+w+5eOPP8b8+fMBwP4vsWTJmmJtHEfuXis4TpSvD/369Uu0O/HEE6298847W/uhhx5K\ntHv66aet/dJLL1l7n332sTbHOAPJ+Gq+VlQaN9weHJN6yCGHVLQtj1ai9Ck//vGPE69vueUWa3OV\nPXeMwGMTHge454njhfl676ax5e2zv2K9cspKl2r4ON84yI099sUo+57t8hGTT4lyoKzUHs8UxlJj\nTFul2xaREwG0Adi3vbZK3NRSJ4BqpZVQn6KEUkYr6lOUtYjFp+hAOYdkfJr0TQD800nf4nvutg8E\n8BMA+xpj/BUWlOip4Klj1UrOUJ+ihKJaUUKISSdRDpT/P3vvHn9VXeX/PxekiTcUUERQwcQLmoLi\n/UapDTqTOKNjYqVOllONTpbjN5pGp6yZ6usvmy5eQjMtS0dTv1KRaCZiXgFFFAhEEEVRRBTDO/r+\n/XH22az95uzzeX/O9X3OXs/H4/P4rHP2e+/9Pme/ztrvvd9rr6WrAPmpfEJDIk466aTUfu2111Jb\np4rz95UXKK7b+PvVtn9QdShH//79U3vs2PyLIT3d5/fH
```
Comment (rxwei) [inline S4TF_Tutorial_4.ipynb:1034]:
```suggestion
        "let thirdTrainLabel = Int(y[2][0].scalarized())\n",
```

Comment (rxwei) [inline S4TF_Tutorial_4.ipynb:1176]: Please put `{` at the end of the previous line.

**Review by rxwei — APPROVED** [review]
Nice work!

**Comment by Ayush517** [comment]
@rxwei Everything is same as Tutorial 3, any more changes here?
